### ISSUE TITLE
Needless lifetime false positive with async function.

### ISSUE THREAD

**Issue by jward-bw** [rust-lang/rust-clippy#7271]
Lint name: `needless_lifetimes`

I tried this code:

```rust
      fn add<'a>       (x: &'a u8, _: &u8) -> &'a u8 {x}
async fn async_add<'a> (x: &'a u8, _: &u8) -> &'a u8 {x}
```

I expected to see this happen: 
I don't expect to see any `needless_lifetimes` lint errors. In both functions the `'a` lifetime is required to tell the compiler which input the output borrows from.

Instead, this happened: 
I receive a `needless_lifetimes` error for `async_add`. There isn't a way to resolve this line without ignoring the lint rule, since replacing any combination of the lifetime parameters with the anonymous lifetime result in a compiler error.

### Meta

- `cargo clippy -V`: `clippy 0.1.52 (9bc8c42b 2021-05-09)`
- `rustc -Vv`:
  ```
  rustc 1.52.1 (9bc8c42bb 2021-05-09)
  binary: rustc
  commit-hash: 9bc8c42bb2f19e745a63f3445f1ac248fb015e53
  commit-date: 2021-05-09 
  host: x86_64-apple-darwin
  release: 1.52.1
  LLVM version: 12.0.0
  ```

**Comment by urkle** [comment]
This also is an issue even if the return doesn't use the lifetime bu a nested struct does.. for example
```rust
pub async fn search<'a>(
                &mut self,
                _base: &str,
                _scope: Scope,
                _filter: &str,
                _attrs: Vec<&'a str>,
            ) -> ldap3::result::Result<SearchResult> {
   Err(LdapError::EndOfStream)
}
```

The compiler requires the 'a be there.

What is interesting is that clippy does not report a failure with this code.
```rust
pub async fn add<'a>(
                &mut self,
                _dn: &str,
                _attrs: Vec<(&'a str, HashSet<&'a str>)>,
            ) -> ldap3::result::Result<LdapResult> {
                Err(LdapError::EndOfStream)
            }
```

clippy 0.1.54 (a178d032 2021-07-26)

rustc 1.54.0 (a178d0322 2021-07-26)
binary: rustc
commit-hash: a178d0322ce20e33eac124758e837cbd80a6f633
commit-date: 2021-07-26
host: x86_64-apple-darwin
release: 1.54.0
LLVM version: 12.0.1

**Comment by kangalioo** [comment]
Perhaps this issue should be added to the known problems in https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes?
